Python/bltinmodule.c

```diff
@@ -867,18 +867,17 @@ builtin_compile_impl(PyObject *module, PyObject *source, PyObject *filename,
         goto error;
 
 #ifdef Py_GIL_DISABLED
-    // gh-118527: Disable immortalization of code constants for explicit
+    // Disable immortalization of code constants for explicit
     // compile() calls to get consistent frozen outputs between the default
     // and free-threaded builds.
-    // Subtract two to suppress immortalization (so that 1 -> -1)
-    PyInterpreterState *interp = _PyInterpreterState_GET();
-    _Py_atomic_add_int(&interp->gc.immortalize, -2);
+    _PyThreadStateImpl *tstate = (_PyThreadStateImpl *)_PyThreadState_GET();
+    tstate->suppress_immortalization++;
```

Comment:
It looks like we were previously using atomics to modify the equivalent counter. Why don't we need to do that here? Same question for the inc/dec below.

Reply:
Previously, the suppression was per-interpreter instead of per-thread. We wanted per-interpreter behavior because it made it easier to apply `@suppress_immortalization()` to tests that internally used threads. Atomics were needed because concurrent calls to `suppress_immortalization()`  might race.

We now only suppress immortalization around `compile` and `eval` calls, which don't spawn threads internally so it's simpler and easier to keep track of this per-thread. And since it's per-thread, we don't need atomics.